### PR TITLE
read_html(): rewinding [wip]

### LINKED ISSUE #17975: BUG: error in read_html when parsing badly-escaped HTML from an io object
#### Code Sample, a copy-pastable example if possible

Create `test.html`, with the contents:
```html
<!doctype html>
<html>
<body>
<table>
	<tr><td>poorly-escaped cell with an & oh noes</td></tr>
</table>
</body>
</html>
```

```py
>>> import pandas as pd
>>> pandas.__version__
'0.20.3'
>>> f = open('./test.html')
>>> pd.read_html(f)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    pd.read_html(f)
  File "/usr/lib/python3.6/site-packages/pandas/io/html.py", line 906, in read_html
    keep_default_na=keep_default_na)
  File "/usr/lib/python3.6/site-packages/pandas/io/html.py", line 743, in _parse
    raise_with_traceback(retained)
  File "/usr/lib/python3.6/site-packages/pandas/compat/__init__.py", line 344, in raise_with_traceback
    raise exc.with_traceback(traceback)
ValueError: No text parsed from document: <_io.TextIOWrapper name='/home/liam/test.html' mode='r' encoding='UTF-8'>
```
#### Problem description

Pandas attempts to invoke a series of parsers on HTML documents, returning when one produces a result, and continuing to the next on error. This works fine when passing a path or entire document to `read_html()`, but when an IO object is passed, the subsequent parsers will be reading from a file whose read cursor is at EOF, producing an inscrutable 'no text parsed from document' error.

This can easily be fixed by rewinding the file with `seek(0)` before continuing to the next parser (will add PR shortly).

#### Expected Output
```
[                                       0
0  poorly-escaped cell with an & oh noes]
```

#### Output of ``pd.show_versions()``

<details>

```
>>> pd.show_versions()
INSTALLED VERSIONS
------------------
commit: e1dabf37645f0fcabeed1d845a0ada7b32415606
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.6-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0rc1+36.ge1dabf376.dirty
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.6.0
Cython: 0.27.2
numpy: 1.13.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: 4.1.0
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```

</details>

### PR DESCRIPTION
**Pull request by erinzm**
- [x] closes #17975
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
tests!

**Comment by liamim** [comment]
@jreback yep, they're coming :smile:

**Comment by liamim** [comment]
Unfortunately, non-seekable io objects typically still have a stub `.seek()` implementation which raises `io.UnsupportedOperation`. I'll trap that instead.

File: pandas/io/html.py [hunk 1]
```diff
@@ -742,6 +742,19 @@ def _parse(flavor, io, match, attrs, encoding, **kwargs):
         try:
             tables = p.parse_tables()
         except Exception as caught:
+            # if `io` is an io-like object, check if it's seekable
+            # and try to rewind it before trying the next parser
+            if hasattr(io, 'seekable') and io.seekable():
+                io.seek(0)
+
```
Comment (jreback) [inline pandas/io/html.py]: make an if else

**Comment by liamim** [comment]
How should I test rewinding on file objects? Just add a malformed HTML file to `pandas/tests/io/data`?

**Comment by jreback** [comment]
there are several examples in the issue
you can also create a class that is has a read and seek method that wraps a url

File: pandas/io/html.py [hunk 2]
```diff
@@ -742,6 +742,18 @@ def _parse(flavor, io, match, attrs, encoding, **kwargs):
         try:
             tables = p.parse_tables()
         except Exception as caught:
+            # if `io` is an io-like object, check if it's seekable
+            # and try to rewind it before trying the next parser
+            if hasattr(io, 'seekable') and io.seekable():
+                io.seek(0)
+            elif hasattr(io, 'seekable') and not io.seekable():
+                # if we couldn't rewind it, let the user know
+                raise ValueError('The favor {} failed to parse your input. '
```
Comment (jschendel) [inline pandas/io/html.py]: favor -> flavor

**Comment by jreback** [comment]
thanks @liamim nice PR!

**Pull request merged**